### PR TITLE
feat: Add Azure Front Door CDN resource

### LINKED ISSUE #382: Azure Front Doors CDNs
### Describe the resource.

Add support for Azure CDN resource

### Use Case

understand configuration in org

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by candiduslynx**
#### Summary

Added Azure Front Door CDN resource (#382).

---

Use the following steps to ensure your PR is ready to be reviewed

- [x] Read the [contribution guidelines](../blob/main/CONTRIBUTING.md) 🧑‍🎓
- [x] Run `go fmt` to format your code 🖊
- [x] Lint your changes via `golangci-lint run` 🚨 (install golangci-lint [here](https://golangci-lint.run/usage/install/#local-installation))
- [x] Update or add tests. Learn more about testing [here](https://docs.cloudquery.io/docs/developers/sdk/testing) 🧪
- [x] Update the docs by running `go run ./docs/docs.go` and committing the changes 📃
- [x] If adding a new resource, add [relevant Terraform files](../terraform) in a separate PR 📂
- [x] Ensure the status checks below are successful ✅

### CODE REVIEW

File: docs/contributing/adding_a_new_resource.md [hunk 1]
```diff
@@ -44,7 +45,7 @@ For complex fields or fields that require further API calls, you can defined you
 
 #### Implementing Resolver Functions
 
-A few important things to note when adding functions that call the AWS API:
+A few important things to note when adding functions that call the Azure API:
```
Comment (hermanschaaf) [inline docs/contributing/adding_a_new_resource.md:48]: Thanks for all these fixes 🙌

File: resources/services/frontdoor/front_doors_mock_util_test.go [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+import (
+	"math/rand"
+	"testing"
+
+	"github.com/Azure/azure-sdk-for-go/services/frontdoor/mgmt/2020-11-01/frontdoor"
+	"github.com/cloudquery/faker/v3"
+)
+
+func fakeBasicRouteConfiguration(t *testing.T) frontdoor.BasicRouteConfiguration {
+	switch rand.Int() % 3 {
```
Comment (hermanschaaf) [inline resources/services/frontdoor/front_doors_mock_util_test.go:12]: I'm a bit wary of using the `rand` package directly in tests - we don't know if something else might be changing the seed and therefore causing the test to run non-deterministically. I see two options:
 1. Instantiate a new `Rand` object and use it throughout, so we know the seed is fixed, e.g. `r := rand.New(rand.NewSource(1))`
 2. Keep track of a counter to do the modulo operation on, and increment it each time this function is called.
 
My preference would lean slightly towards #1, but either one is fine imo, as long as there's no chance of the test behavior changing randomly either now or in the future if the seed gets changed somewhere else.

File: resources/services/frontdoor/front_doors_mock_util_test.go [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+	if err := faker.FakeDataSkipFields(&backendPoolsSettings, []string{"EnforceCertificateNameCheck"}); err != nil {
+		t.Fatal(err)
+	}
+	backendPoolsSettings.EnforceCertificateNameCheck = frontdoor.EnforceCertificateNameCheckEnabledStateEnabled
+	properties.BackendPoolsSettings = &backendPoolsSettings
+	properties.EnabledState = frontdoor.EnabledStateEnabled
+
+	frontDoor.Properties = &properties
+	return frontDoor
+}
```
Comment (hermanschaaf) [inline resources/services/frontdoor/front_doors_mock_util_test.go:289]: Wow, this was a lot of work to fake 🤯  Nice work!

File: resources/services/frontdoor/gen.hcl [hunk 4]
```diff
@@ -0,0 +1,561 @@
...
+    }
+  }
+
+  relation "azure" "frontdoor" "properties_load_balancing_settings" {
+    rename      = "load_balancing_settings"
+    description = "Load balancing settings for a backend pool associated with the Front Door instance"
+
+    column "load_balancing_settings_properties_resource_state" {
+      rename      = "resource_state"
+      description = "Resource status"
```
Comment (hermanschaaf) [inline resources/services/frontdoor/gen.hcl:282]:
```suggestion
      description = "Resource state"
```

File: docs/tables/azure_front_door_health_probe_settings.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
+
+# Table: azure_front_door_health_probe_settings
+Health probe settings for a backend pool associated with this Front Door instance
+## Columns
+| Name        | Type           | Description  |
+| ------------- | ------------- | -----  |
+|front_door_cq_id|uuid|Unique CloudQuery ID of azure_front_doors table (FK)|
+|resource_state|text|Resource status|
+|path|text|The path to use for the health probe|
+|health_probe_settings_properties_protocol|text|Protocol - Protocol scheme to use for this probe|
```
Comment (hermanschaaf) [inline docs/tables/azure_front_door_health_probe_settings.md:10]: This should probably be renamed `protocol`

File: resources/services/frontdoor/gen.hcl [hunk 6]
```diff
@@ -0,0 +1,561 @@
...
+    }
+
+    column "health_probe_settings_properties_path" {
+      rename      = "path"
+      description = "The path to use for the health probe"
+    }
+
+    column "protocol" {
```
Comment (hermanschaaf) [inline resources/services/frontdoor/gen.hcl:327]: Referring to a comment I made higher up, I think this should be `health_probe_settings_properties_protocol`. 

On a side note (not for this PR), I think we can make some improvements to cq-gen here:
 - It should warn you when you've defined a column that it didn't find in the source. I started implementing this but got stuck at some point; I might try again when I'm more familiar with the source.
 - It seems clunky that it requires you to specify these long column names when `health_probe_settings` is already in the table name--seems like that prefix shouldn't be necessary, but maybe I'm missing something?

**Review by hermanschaaf — CHANGES_REQUESTED** [review]
Great stuff! :tada: I commented on a couple of small things I noticed while looking through, but this is an amazing addition.

File: resources/services/frontdoor/front_doors_mock_util_test.go (hunk 2 above)
Comment (yevgenypats) [reply resources/services/frontdoor/front_doors_mock_util_test.go:12]: Good catch. Agreed. better not to use any random stuff in tests

File: resources/services/frontdoor/gen.hcl (hunk 6 above)
Comment (candiduslynx) [reply resources/services/frontdoor/gen.hcl:327]: Yes, the 2nd option would be really great (basically, I think `skip_prefix` should work not only for embedded relations, but for straightforward stuff as well. Or the generator shouldn't add prefix by itself).

**Review by hermanschaaf — APPROVED** [review]

File: resources/services/frontdoor/front_doors.go [hunk 7]
```diff
@@ -0,0 +1,910 @@
...
+								Type:        schema.TypeJSON,
+								Resolver:    resolveFrontDoorRulesEngineRulesRouteConfigurationOverride,
+							},
+							{
+								Name:        "match_processing_behavior",
+								Description: "If the rule is a match should the rules engine continue running the remaining rules or stop",
+								Type:        schema.TypeString,
+							},
+						},
+						Relations: []*schema.Table{
```
Comment (roneli) [inline resources/services/frontdoor/front_doors.go:174]: maybe make actions as a json?

File: resources/services/frontdoor/front_doors.go [hunk 8]
```diff
@@ -0,0 +1,910 @@
...
+						Type:        schema.TypeString,
+					},
+					{
+						Name:        "id",
+						Description: "Resource ID",
+						Type:        schema.TypeString,
+						Resolver:    schema.PathResolver("ID"),
+					},
+				},
+				Relations: []*schema.Table{
```
Comment (roneli) [inline resources/services/frontdoor/front_doors.go:357]: maybe an array of strings? instead of a relation?

**Review by roneli — APPROVED** [review]
LGTM, had a few nits.

File: resources/services/frontdoor/front_doors.go (hunk 8 above)
Comment (candiduslynx) [reply resources/services/frontdoor/front_doors.go:357]: done

File: resources/services/frontdoor/front_doors.go (hunk 7 above)
Comment (candiduslynx) [reply resources/services/frontdoor/front_doors.go:186]: done

**Review by roneli — APPROVED** [review]
LGTM, well done!

**Pull request merged**